src/nodes.coffee

```diff
@@ -1894,6 +1897,48 @@ exports.Obj = class Obj extends Base
       answer.push @makeCode join
     if @front then @wrapInParentheses answer else answer
 
+  expandProperty: (property) ->
+    {variable, context, operatorToken} = property
+    key = if property instanceof Assign and context is 'object'
+      variable
+    else if property instanceof Assign
+      operatorToken.error "unexpected #{operatorToken.value}" unless @lhs
+      variable
+    else
+      property
+    if key instanceof Value and key.hasProperties()
+      key.error 'invalid object key' unless context isnt 'object' and key.this
+      return new Assign(key, property, 'object', shorthand: yes).withLocationDataFrom property
+    return property unless key is property
+    return property if property instanceof Splat
+
+    new Assign(property, property, 'object', shorthand: yes).withLocationDataFrom property
+
+  expandProperties: ->
+    @expandProperty(property) for property in @properties
+
+  getPropertyAst: (property) ->
+    return property.ast() if property instanceof Splat
+
+    {variable, value, shorthand} = property
+    isComputedPropertyName = variable instanceof Value and variable.base instanceof ComputedPropertyName
+
+    Object.assign
+      type: 'ObjectProperty'
+      key: variable.ast()
+      value: value.ast()
+      shorthand: !!shorthand
+      computed: !!isComputedPropertyName
+      method: no
+    , property.astLocationData()
+
+  astType: -> 'ObjectExpression'
```

Comment:
As with arrays, this will get more logic once we have ASTs for assignments (and thus left-hand-side objects)